ISPN-13489 Exponential back-off does not work with multiple sites

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-13489

I've created `CompletableFuture`, which can be retried multiple times. We have the back-off for the whole `DefaultIracManager#run` and a back-off per operation issued to the remote sites. Both follow the same retry policy.

At first, I was trying to use the same thread for the retries with the `WithinThreadExecutor`, but it started to block too much stuff. Because of the way I've created the `RetryableCompletionStage`, where the `onComplete` calls the `retry`, which calls the `onComplete`. So suggestions are really welcome.

pruivo commented:
At the first glance, I don't think this fixes the issue at all. It looks like if one site is on back-off, it blocks all updates to other sites too. It is not what is intended... However, I need to test it to confirm.

pruivo commented:
As an alternative solution that I had in mind when I created the JIRA:

Just a flag per backup site telling if back-off for that site is enabled or not.

On network exception do
* set flag 
* set the timer to reset the flag

Line https://github.com/infinispan/infinispan/blob/ba1c688839792a9b1519f2eb8c6fc75d0421bb7a/core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java#L441, do not send anything for sites that are in back-off

On IracManagerKeyState, add state to tell which sites haven't seen the update yet (bonus, allows to fix another issue with state transfer sending everything to all sites).

jabolina commented:
@pruivo  Just updated what I did before following your comment. I'll add a few comments to the code too with my reasoning.

Updated to:

* Sites can be but in back-off independently;
* Some sites can be in back-off while the manager can still send requests to others;

This added the complexity that some sites might see some states while others don't. To tackle this I did:

* We keep track of which sites received each state;
* Once a site has received a state, we do not send it again.
* Each site might receive different commands with the states, updated `IracResponseCollector` to handle a request to a single backup;
* State is done/removed *only* after all backups received them.

I did not touch the clear part though. With the back-off removed from the executor, we could drown the other sites here?

jabolina commented on core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:
```diff
@@ -182,9 +183,8 @@ public CompletionStage<Void> trackForStateTransfer(Collection<XSiteState> stateL
-         if (updatedKeys.putIfAbsent(iracState.getKey(), iracState) == null) {
-            cf.dependsOn(iracState.getCompletionStage());
-         }
+         IracManagerKeyState existent = updatedKeys.computeIfAbsent(iracState.getKey(), ignore -> iracState);
+         cf.dependsOn(existent.subscribe());
```
Found this during the `IracStateTransferTest#testConflict`. The test will disconnect the sites, put some values in each, and connect them again, waiting for the conflict resolution to take place. I found that:

1. The site will have some state to send to the backup, since the backup is offline we skip sending the command;
2. The manager invoked `IracResponseCollector#freeze`, which in turn calls the `IracResponseCollector#run`

Since the site was offline, no data was sent to the backup. But because the `IracResponseCollector` was created passing the batch of states and when `IracResponseCollector#run` executes it will not find anything in the collector's `failedKeys` set, and then call the `DefaultIracManager#onBatchResponse` saying that all states were successfully sent, and then clearing all states tracked. Even though nothing was sent to the backup because it was offline.

When the test calls for the state transfer, it will add the `IracManagerStateTransferState` to `updatedKeys` and will wait for it to complete/transfer. The test then is able to see that everything is transferred ok.

When I inserted my changes, we did not remove the elements from the `updatedKeys` because we were unable to send it to the backup because it was offline. So the state transfer can't insert the `IracManagerStateTransferState`, because the state is still tracked, which would cause to return saying that the state transfer was finished (the `iracExecutor.run()` in the next line is asynchronous) even though no state was transferred yet. I added the completable future to the `IracManagerKeyState` just so we are able to wait here for the transfer to complete.

pruivo replied:
> When I inserted my changes, we did not remove the elements from the updatedKeys because we were unable to send it to the backup because it was offline.

This is a problem. Users can take a site offline for long periods and, in this case, you don't want to keep keys there wasting memory (and increasing memory consumption).

pruivo commented:
@jabolina needs rebase

jabolina commented:
Multiple tests failed with timeout and CI taking much longer to complete. I'll try to identify something.

jabolina commented:
@pruivo I think that's it. Fixed the last pending tests, but both `XSiteIT` and `XSiteServerTest` are failing in all branches. I'll open a JIRA to fix them.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracBackOffHandler.java:
```diff
@@ -0,0 +1,73 @@
+   private volatile boolean putForBackOff;
+
+   public IracBackOffHandler(Collection<IracXSiteBackup> backups, ExponentialBackOff backOff, Executor executor) {
+      this.backups = Immutables.immutableSetWrap(Set.copyOf(backups));
```
not required since `backups` if fine when this instance is created.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracBackOffHandler.java:
```diff
@@ -0,0 +1,73 @@
+      if (putForBackOff) {
+         //noinspection NonAtomicOperationOnVolatileField
+         runnable = runnable.thenCompose(ignore -> backOff.asyncBackOff())
+               .thenComposeAsync(this, executor);
```
`apply()` always return a completed future. This class can implement `Runnable` and here you can invoke `thenRunAsync()`.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracBackOffHandler.java:
```diff
@@ -0,0 +1,73 @@
+   }
+
+   private CompletableFuture<Void> handleBackOff() {
+      if (putForBackOff) {
```
I'm not sure why you need the boolean. `enableBackOff` can inline the code inside this `if` branch and `disableBackOff` invoke the `else` branch.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracBackOffHandler.java:
```diff
@@ -0,0 +1,73 @@
+      this.runnable = CompletableFutures.completedNull();
+   }
+
+   public CompletableFuture<Void> enableBackOff() {
```
this method could receive the runnable as a parameter to be invoked when the back-off ends.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracManagerKeyChangedState.java:
```diff
@@ -92,6 +103,27 @@ public void discard() {
+   }
+
+   @Override
+   public Collection<XSiteBackup> successfullySent() {
```
Only used to check for contains all. `containsAll` can be invoked directly here and avoid copying.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:
```diff
@@ -557,23 +578,34 @@ private void onBatchResponse(IracBatchSendResult result, Collection<? extends Ir
             break;
          case RETRY:
-            iracExecutor.disableBackOff();
+            backOffHandler.disableBackOff();
```
I don't think you need the method `disableBackOff()`. The response collector can invoke `org.infinispan.xsite.XSiteBackup#disableBackOffMode` or `org.infinispan.xsite.XSiteBackup#enableBackOffMode`.

jabolina replied:
I've just kept a call to reset the backoff interval.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java:
```diff
@@ -1,136 +1,111 @@
+            backup.enableBackOffMode();
+            result = IracBatchSendResult.BACK_OFF_AND_RETRY;
+         } else if (result == IracBatchSendResult.OK) {
+            //don't overwrite communication errors
```
as mentioned in the previous comment, you can invoke `backup.disableBackOffMode();` here.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java:
```diff
@@ -1,136 +1,111 @@
+      } else {
+         if (trace) {
+            log.tracef("[IRAC] Received response from site %s: %s", backup.getSiteName(), rspIntSet);
+         }
```
and here too

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java:
```diff
@@ -1,136 +1,111 @@
-         completableFuture.complete(null);
-         return;
-      }
+   private void run() {
```
wouldn't mind to copy the method code to the above method and remove  `run`, which means nothing at this point :)

pruivo reviewed (changes requested):
Overall, looks good to me. Just have some minor suggestion here and there :+1:

jabolina commented:
Thanks for the review, applied all suggestions.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:
```diff
@@ -557,23 +577,33 @@ private void onBatchResponse(IracBatchSendResult result, Collection<? extends Ir
       switch (result) {
          case OK:
-            iracExecutor.disableBackOff();
+            backOffHandler.resetBackOff();
```
this is not working. 
```
2023-05-22 18:59:34,491 ERROR (blocking-thread--p3-t3) [o.j.p.r.RELAY2] JGRP000290: pedro-desktop-14560:site1: no route to site site3: dropping message (89 identical messages for site site3 in the last 60629 ms)
2023-05-22 19:00:34,646 ERROR (blocking-thread--p3-t3) [o.j.p.r.RELAY2] JGRP000290: pedro-desktop-14560:site1: no route to site site3: dropping message (85 identical messages for site site3 in the last 60155 ms)
2023-05-22 19:01:35,197 ERROR (blocking-thread--p3-t3) [o.j.p.r.RELAY2] JGRP000290: pedro-desktop-14560:site1: no route to site site3: dropping message (86 identical messages for site site3 in the last 60551 ms)
```

I think each site has to have its own `org.infinispan.util.ExponentialBackOff` in `org.infinispan.xsite.irac.IracXSiteBackup`, so the `reset` only resets the timer for that site.

jabolina replied:
Updated, now each site has its own backoff timing.

pruivo commented on core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:
```diff
@@ -417,56 +419,72 @@ private CompletableSource sendUpdateBatch(Collection<? extends IracStateData> ba
+
+      AggregateCompletionStage<Void> aggregation = CompletionStages.aggregateCompletionStage();
+      for (IracXSiteBackup backup: asyncBackups) {
+         IracPutManyCommand cmd = commandsFactory.buildIracPutManyCommand(size);
```
you can "fail" fast in 2 cases

* if `backup.isBackOffMode()` -> continue // there is nothing to do
* if backup is offline -> for each data in batch do `data.successFor(backup)` // otherwise, state is never removed.

jabolina replied:
Updated for the `backup.isBackOffMode()`. For the offline, I kept getting some test failures.

pruivo commented:
looks like there are still tests failing

jabolina commented:
It was a fix on the test only. Between removing the exception to throw and verifying, the executor could be invoked and add more events to the queue.

pruivo commented:
merged! thanks @jabolina !
